src/node_native_module_env.cc

```diff
@@ -38,6 +43,61 @@ Local<String> NativeModuleEnv::GetConfigString(Isolate* isolate) {
   return NativeModuleLoader::GetInstance()->GetConfigString(isolate);
 }
 
+bool NativeModuleEnv::CompileAllModules(Local<Context> context) {
+  NativeModuleLoader* loader = NativeModuleLoader::GetInstance();
+  std::vector<std::string> ids = loader->GetModuleIds();
+  bool all_succeeded = true;
+  for (const auto& id : ids) {
+    // TODO(joyeecheung): compile non-module scripts here too.
+    if (!loader->CanBeRequired(id.c_str())) {
+      continue;
+    }
+    v8::TryCatch bootstrapCatch(context->GetIsolate());
+    native_module::NativeModuleLoader::Result result;
+    USE(loader->CompileAsModule(context, id.c_str(), &result));
+    if (bootstrapCatch.HasCaught()) {
+      per_process::Debug(DebugCategory::CODE_CACHE,
+                         "Failed to compile code cache for %s\n",
+                         id.c_str());
+      all_succeeded = false;
+      PrintCaughtException(context->GetIsolate(), context, bootstrapCatch);
+    }
+  }
+  return all_succeeded;
+}
+
+void NativeModuleEnv::CopyCodeCache(std::vector<CodeCacheInfo>* out) {
+  NativeModuleLoader* loader = NativeModuleLoader::GetInstance();
+  Mutex::ScopedLock lock(loader->code_cache_mutex());
+  auto in = loader->code_cache();
+  for (auto const& item : *in) {
+    out->push_back(
+        {item.first,
+         {item.second->data, item.second->data + item.second->length}});
```

Comment:
Here the data is not been copied -- not as the method name suggests, thus the out pointer may become a dangling pointer. Though the use case in the snapshot builder doesn't seem to have any problem with it, I think it would be better to align the behavior and the naming.

Reply:
I think it is copied, since it's std::vector?

Reply:
Oh, you are right. Sorry about this one.